File: apple/HavenApp/BackupDetail.swift
import SwiftUI

/// "Where is this post actually stored?" — tap the cloud indicator to find out.
///
/// The indicator answers yes/no. That was enough right up until the day it said yes and nobody could
/// fetch anything: every blob had reached a relay, the relay was the one running inside the author's
/// own app, and the tick had no way to say so. The only way to learn which relay held what was to
/// read the app's log stream. This is that answer, in the app.
///
/// Grouped by RELAY, not by attachment. The per-attachment shape repeated an identical block (and an
/// identical warning) once per photo — four photos meant four screens of the same sentence, and the
/// one fact that matters, "which relay is missing copies", had to be reassembled by eye. A count per
/// relay says the same thing in one line and still names a partial failure exactly ("3 of 4").
struct BackupRefs: Identifiable {
    let refs: [String]
    var id: String { refs.joined(separator: "|") }
}

struct BackupDetailView: View {
    let refs: [String]
    let circleId: String
    @ObservedObject private var relayStore = RelayMailboxStore.shared
    @Environment(\.dismiss) private var dismiss

    private var ownRelayHex: String { RelayHost.shared.serving ? RelayHost.shared.nodeId : "" }

    /// Every relay worth a row: the ones this circle publishes to, plus any that already hold a copy.
    /// Naming a circle relay holding NOTHING is the case you most need to see, and a list of
    /// confirmations alone can never show it.
    private var rows: [(dest: String, have: Int)] {
        var dests = Set(relayStore.relays(forCircle: circleId))
        for ref in refs { dests.formUnion(MediaBackupLedger.destinations(for: ref)) }
        return dests.sorted().map { dest in
            (dest, refs.filter { MediaBackupLedger.destinations(for: $0).contains(dest) }.count)
        }
    }

    /// True when nothing but our own in-process relay holds a full set — the state where the post
    /// looks backed up and is in fact unreachable to everyone else.
    private var strandedOnOwnRelay: Bool {
        let remoteComplete = rows.contains { $0.dest != ownRelayHex && $0.have == refs.count }
        let ownHasAny = rows.contains { $0.dest == ownRelayHex && $0.have > 0 }
        return ownHasAny && !remoteComplete
    }

    var body: some View {
        NavigationStack {
            List {
                Section {
                    ForEach(rows, id: \.dest) { row in
                        relayRow(dest: row.dest, have: row.have)
                    }
                    if rows.isEmpty {
                        Label("Not on any relay yet", systemImage: "xmark.circle")
                            .foregroundStyle(.secondary).font(.callout)
                    }
                } header: {
                    Text(refs.count == 1 ? "1 attachment" : "\(refs.count) attachments")
                } footer: {
                    if strandedOnOwnRelay {
                        Text("Only this device's own relay has a copy, so nobody else can fetch it yet.")
                            .foregroundStyle(.orange)
                    }
                }
            }
            .navigationTitle("Where this is stored")
            #if os(iOS)
            .navigationBarTitleDisplayMode(.inline)
            #endif
            .toolbar { ToolbarItem(placement: .confirmationAction) { Button("Done") { dismiss() } } }
        }
    }

    @ViewBuilder private func relayRow(dest: String, have: Int) -> some View {
        let isOwn = dest == ownRelayHex
        let all = have == refs.count
        HStack(spacing: 10) {
            Image(systemName: have == 0 ? "circle.dotted"
                            : (isOwn ? "externaldrive.badge.exclamationmark"
                                     : (all ? "checkmark.circle.fill" : "circle.lefthalf.filled")))
                .foregroundStyle(have == 0 ? Color.secondary : (isOwn ? Color.orange : (all ? Color.green : Color.orange)))
            Text(name(for: dest)).font(.callout)
            Spacer()
            Text(statusText(have: have, isOwn: isOwn))
                .font(.caption).foregroundStyle(.secondary)
        }
    }

    private func statusText(have: Int, isOwn: Bool) -> String {
        if have == 0 { return "No copy yet" }
        let count = have == refs.count ? "All" : "\(have) of \(refs.count)"
        return isOwn ? "\(count) · on this device" : count
    }

    /// A relay's friendly name, falling back to a short hex. An S3 destination is stored by bucket
    /// rather than node hex, so it is already readable.
    private func name(for dest: String) -> String {
        if let e = relayStore.entries[dest] { return e.name }
        if dest.contains("/") || dest.contains(".") { return dest }   // s3-style destination
        return "Relay · \(dest.prefix(8))…"
    }
}

File: android/app/src/main/java/com/blaineam/haven/ui/BackupDetail.kt
package com.blaineam.haven.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.RadioButtonUnchecked
import androidx.compose.material.icons.filled.Smartphone
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import com.blaineam.haven.R
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.blaineam.haven.core.HavenNet

/**
 * "Where is this post actually stored?" — tap the cloud indicator to find out. Apple parity
 * (`BackupDetailView`); Android and desktop had no equivalent at all, so the only way to learn which
 * relay held what on this platform was to read logcat.
 *
 * The indicator answers yes/no. That was enough right up until the day it said yes and nobody could
 * fetch anything: every blob had reached a relay, the relay was the one running inside the author's
 * own app, and the tick had no way to say so.
 *
 * Grouped by RELAY, not by attachment: the per-attachment shape repeats an identical block once per
 * photo, and the one fact that matters — which relay is missing copies — has to be reassembled by
 * eye. A count per relay says the same thing in one line and still names a partial failure exactly
 * ("3 of 4").
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun BackupDetailSheet(refs: List<String>, circleId: String, onDismiss: () -> Unit) {
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = false)
    val ownRelayHex = remember { HavenNet.ownHostedRelayHex() }

    // Every relay worth a row: the ones this circle publishes to, PLUS any that already hold a copy.
    // Naming a circle relay that holds nothing is the case you most need to see.
    val rows = remember(refs, circleId) {
        val dests = LinkedHashSet<String>()
        dests.addAll(HavenNet.circleRelayHexes(circleId))
        refs.forEach { dests.addAll(HavenNet.mediaBackupDestinations(it)) }
        dests.sorted().map { dest ->
            dest to refs.count { HavenNet.mediaBackupDestinations(it).contains(dest) }
        }
    }

    /** Nothing but our own in-process relay holds a full set — it looks backed up and is in fact
     *  unreachable to everyone else. */
    val strandedOnOwnRelay = remember(rows) {
        val remoteComplete = rows.any { it.first != ownRelayHex && it.second == refs.size }
        val ownHasAny = rows.any { it.first == ownRelayHex && it.second > 0 }
        ownHasAny && !remoteComplete
    }

    ModalBottomSheet(onDismissRequest = onDismiss, sheetState = sheetState,
                     containerColor = HavenTheme.card) {
        Column(Modifier.fillMaxWidth().padding(horizontal = 20.dp).padding(bottom = 28.dp)) {
            Text(stringResource(R.string.backup_title),
                color = HavenTheme.textPrimary, fontSize = 18.sp, fontWeight = FontWeight.SemiBold)
            Spacer(Modifier.height(4.dp))
            Text(if (refs.size == 1) stringResource(R.string.backup_one_attachment) else stringResource(R.string.backup_n_attachments, refs.size),
                color = HavenTheme.textSecondary, fontSize = 12.sp)
            Spacer(Modifier.height(14.dp))

            if (rows.isEmpty()) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Icon(Icons.Filled.RadioButtonUnchecked, null,
                        tint = HavenTheme.textSecondary, modifier = Modifier.size(18.dp))
                    Spacer(Modifier.size(10.dp))
                    Text(stringResource(R.string.backup_no_relay), color = HavenTheme.textSecondary, fontSize = 14.sp)
                }
            }
            rows.forEach { (dest, have) ->
                val isOwn = dest == ownRelayHex
                val all = have == refs.size
                Row(
                    Modifier.fillMaxWidth().padding(vertical = 7.dp),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.SpaceBetween,
                ) {
                    Row(verticalAlignment = Alignment.CenterVertically, modifier = Modifier.weight(1f)) {
                        Icon(
                            when {
                                have == 0 -> Icons.Filled.RadioButtonUnchecked
                                isOwn -> Icons.Filled.Smartphone
                                all -> Icons.Filled.CheckCircle
                                else -> Icons.Filled.Warning
                            },
                            null,
                            tint = when {
                                have == 0 -> HavenTheme.textSecondary
                                isOwn -> Color(0xFFF59E0B)
                                all -> Color(0xFF22C55E)
                                else -> Color(0xFFF59E0B)
                            },
                            modifier = Modifier.size(18.dp),
                        )
                        Spacer(Modifier.size(10.dp))
                        Text(relayLabel(dest), color = HavenTheme.textPrimary, fontSize = 14.sp, maxLines = 1)
                    }
                    Text(
                        statusText(have, refs.size, isOwn),
                        color = HavenTheme.textSecondary, fontSize = 12.sp,
                    )
                }
            }

            if (strandedOnOwnRelay) {
                Spacer(Modifier.height(10.dp))
                Box(Modifier.fillMaxWidth().background(Color(0x22F59E0B)).padding(10.dp)) {
                    Text(stringResource(R.string.backup_stranded),
                        color = Color(0xFFF59E0B), fontSize = 12.sp)
                }
            }
        }
    }
}

@Composable
private fun statusText(have: Int, total: Int, isOwn: Boolean): String {
    if (have == 0) return stringResource(R.string.backup_no_copy)
    val count = if (have == total) stringResource(R.string.backup_all) else stringResource(R.string.backup_x_of_y, have, total)
    return if (isOwn) stringResource(R.string.backup_on_this_device, count) else count
}

/** A relay's friendly name, falling back to a short hex. An S3 destination is stored by bucket
 *  rather than node hex, so it is already readable. */
@Composable
private fun relayLabel(dest: String): String {
    val known = runCatching { HavenNet.relayName(dest) }.getOrNull()
    if (!known.isNullOrBlank()) return known
    if (dest.contains("/") || dest.contains(".")) return dest   // s3-style destination
    return stringResource(R.string.backup_relay_short, dest.take(8))
}
